packages/wonder-blocks-tooltip/components/tooltip-bubble.js

```diff
@@ -1,77 +1,69 @@
 // @flow
-import {css, StyleSheet} from "aphrodite";
+import {StyleSheet} from "aphrodite";
 import * as React from "react";
-import {Popper} from "react-popper";
+import * as ReactDOM from "react-dom";
 
 import {View} from "@khanacademy/wonder-blocks-core";
 import TooltipContent from "./tooltip-content.js";
 import TooltipArrow from "./tooltip-arrow.js";
-import visibilityModifierDefaultConfig from "../util/visibility-modifier.js";
 
-import type {PopperChildrenProps} from "react-popper";
-import type {Placement} from "../util/types.js";
+import type {PopperChildrenProps} from "./tooltip-popper.js";
 
 type Props = {|
     // The content to be shown in the bubble.
     children: React.Element<typeof TooltipContent>,
 
-    // The element that anchors the tooltip bubble.
-    // This is used to position the bubble.
-    anchorElement: ?HTMLElement,
-
-    // Where should the bubble try to go with respect to its anchor.
-    placement: Placement,
+    // The props provided by TooltipPopper, which are used to position
+    // and manage the bubble contents.
+    popperProps: PopperChildrenProps,
 |};
 
 export default class TooltipBubble extends React.Component<Props> {
-    _renderBubble(childrenProps: PopperChildrenProps) {
-        const {children, placement} = this.props;
-        const {ref, outOfBoundaries, style, arrowProps} = childrenProps;
-        const actualPlacement = childrenProps.placement || placement;
+    _lastRef: ?HTMLElement;
 
-        // Here we take the props provided to us by popper.js and use them
-        // to position ourselves. We don't use a View because the ref callback
-        // for the react-popper integration expects an HTMLElement and coercing
-        // a View ref to the right type is a lot of extra lifting for little
-        // gain over just using a div.
-        return (
-            <div
-                data-placement={actualPlacement}
-                ref={ref}
-                className={css(styles.bubble, outOfBoundaries && styles.hide)}
-                style={style}
-            >
-                <View style={styles[`content-${actualPlacement}`]}>
-                    {children}
-                    <TooltipArrow
-                        placement={actualPlacement}
-                        popperArrowProps={arrowProps}
-                    />
-                </View>
-            </div>
-        );
+    _updateRef(ref: ?(React.Component<*> | Element)) {
+        const {popperProps} = this.props;
+        // We only want to update the popper's arrow reference if it is
+        // actually changed. Otherwise, we end up in an endless loop of updates
+        // as every render would trigger yet another render.
+        if (popperProps && ref) {
+            const domNode = ReactDOM.findDOMNode(ref);
+            if (domNode instanceof HTMLElement && domNode !== this._lastRef) {
+                this._lastRef = domNode;
+                popperProps.ref(domNode);
+            }
+        }
```

Comment:
We changed to use `View`, so now we need to do some extra things. Also, nesting the refs means we have to be cautious not to repeatedly call the popper's ref handlers with the same element, or we'll get in a horrible state.

Reply:
Awesome!  I'm glad there wasn't too much additional work here.